pkg/debug/checkups/osquery.go

```diff
@@ -0,0 +1,147 @@
+package checkups
+
+import (
+	"archive/zip"
+	"context"
+	"fmt"
+	"io"
+	"os/exec"
+	"runtime"
+	"strings"
+	"time"
+)
+
+type osqueryCheckup struct {
+	status         Status
+	executionTimes map[string]string // maps command to how long it took to run, in ms
+	summary        string
+}
+
+func (o *osqueryCheckup) Name() string {
+	return "Osquery"
+}
+
+func (o *osqueryCheckup) Run(ctx context.Context, extraWriter io.Writer) error {
+	// Determine passing status by running osqueryd --version
+	o.executionTimes = make(map[string]string)
+	if osqueryVersion, err := o.version(ctx); err != nil {
+		o.status = Failing
+		o.summary = err.Error()
+		return fmt.Errorf("running osqueryd version: %w", err)
+	} else {
+		o.status = Passing
+		o.summary = osqueryVersion
+	}
+
+	// Run launcher interactive to capture timing details
+	if err := o.interactive(ctx); err != nil {
+		return fmt.Errorf("running launcher interactive: %w", err)
+	}
+
+	// If we're running doctor and not flare, that's all we need
+	if extraWriter == io.Discard {
+		return nil
+	}
+
+	zipWriter := zip.NewWriter(extraWriter)
+	defer zipWriter.Close()
+
+	if err := o.foreground(ctx, zipWriter); err != nil {
+		return fmt.Errorf("running osqueryd in foreground: %w", err)
+	}
+
+	return nil
+}
+
+func (o *osqueryCheckup) version(ctx context.Context) (string, error) {
+	osquery := osquerydPath()
+
+	cmdCtx, cmdCancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cmdCancel()
+
+	cmd := exec.CommandContext(cmdCtx, osquery, "--version")
+	startTime := time.Now().UnixMilli()
+	out, err := cmd.CombinedOutput()
+	o.executionTimes[strings.Join(cmd.Args, " ")] = fmt.Sprintf("%d ms", time.Now().UnixMilli()-startTime)
+	if err != nil {
+		return "", fmt.Errorf("running %s version: err %w, output %s", osquery, err, string(out))
+	}
+
+	return strings.TrimSpace(string(out)), nil
+}
+
+func (o *osqueryCheckup) interactive(ctx context.Context) error {
```

Comment:
this is clever! using interactive to test no only osquery, but also that launcher and osquery can communicate

Reply:
haha. I'd assumed using the underlying libraries, but this is pretty clever.